connection_pool/connection_pool_test.go

```diff
@@ -233,6 +234,251 @@ func TestClose(t *testing.T) {
+	}
+
+	return nil
+}
+
+func TestConnectionHandlerOpenUpdateClose(t *testing.T) {
+	poolServers := []string{servers[0], servers[1]}
+	roles := []bool{false, true}
+
+	err := test_helpers.SetClusterRO(poolServers, connOpts, roles)
+	require.Nilf(t, err, "fail to set roles for cluster")
+
+	h := &testHandler{}
+	poolOpts := connection_pool.OptsPool{
+		CheckTimeout:      100 * time.Microsecond,
+		ConnectionHandler: h,
+	}
+	pool, err := connection_pool.ConnectWithOpts(poolServers, connOpts, poolOpts)
+	require.Nilf(t, err, "failed to connect")
+	require.NotNilf(t, pool, "conn is nil after Connect")
+
+	_, err = pool.Call17("box.cfg", []interface{}{map[string]bool{
+		"read_only": true,
+	}}, connection_pool.RW)
+	require.Nilf(t, err, "failed to make ro")
+
+	for i := 0; i < 100; i++ {
+		// Wait for read_only update, it should report about close connection
+		// with old role.
+		if h.deactivated >= 1 {
+			break
+		}
+		time.Sleep(poolOpts.CheckTimeout)
+	}
+	require.Equalf(t, h.deactivated, 1, "updated not reported as deactivated")
+	require.Equalf(t, h.discovered, 3, "updated not reported as discovered")
+
+	pool.Close()
+
+	for i := 0; i < 100; i++ {
+		// Wait for close of all connections.
+		if h.deactivated >= 3 {
+			break
+		}
+		time.Sleep(poolOpts.CheckTimeout)
+	}
+
+	for _, err := range h.errs {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	connected, err := pool.ConnectedNow(connection_pool.ANY)
+	require.Nilf(t, err, "failed to get connected state")
+	require.Falsef(t, connected, "connection pool still be connected")
+	require.Equalf(t, len(poolServers)+1, h.discovered, "unexpected discovered count")
+	require.Equalf(t, len(poolServers)+1, h.deactivated, "unexpected deactivated count")
+}
+
+type testAddErrorHandler struct {
+	connection_pool.ConnectionHandler
+	discovered, deactivated int
+}
+
+func (h *testAddErrorHandler) Discovered(conn *tarantool.Connection,
+	role connection_pool.Role) error {
+	h.discovered++
+	return fmt.Errorf("any error")
+}
+
+func (h *testAddErrorHandler) Deactivated(conn *tarantool.Connection,
+	role connection_pool.Role) error {
+	h.deactivated++
+	return nil
+}
+
+func TestConnectionHandlerOpenError(t *testing.T) {
+	poolServers := []string{servers[0], servers[1]}
+
+	h := &testAddErrorHandler{}
+	poolOpts := connection_pool.OptsPool{
+		CheckTimeout:      100 * time.Microsecond,
+		ConnectionHandler: h,
+	}
+	connPool, err := connection_pool.ConnectWithOpts(poolServers, connOpts, poolOpts)
+	if err == nil {
+		defer connPool.Close()
+	}
+	require.NotNilf(t, err, "success to connect")
+	require.Equalf(t, 2, h.discovered, "unexpected discovered count")
+	require.Equalf(t, 0, h.deactivated, "unexpected deactivated count")
+}
+
+type testUpdateErrorHandler struct {
+	connection_pool.ConnectionHandler
+	discovered, deactivated int
+	mutex                   sync.Mutex
+}
+
+func (h *testUpdateErrorHandler) Discovered(conn *tarantool.Connection,
+	role connection_pool.Role) error {
+	h.mutex.Lock()
```

Comment:
Yes it is, to protect access to testUpdateErrorHandler.discovered and testUpdateErrorHandler.deactivated.

It is possible to do with atomics instead of the mutex, but it can be more complicated. A client code of the our public API is not required to use mutexes if it does not access shared data itself.